src/projection.jl

```diff
@@ -31,114 +68,269 @@ function Base.show(io::IO, project::ProjectTo{T}) where {T}
     end
 end
 
-# fallback (structs)
-function ProjectTo(x::T) where {T}
-    # Generic fallback for structs, recursively make `ProjectTo`s all their fields
+# Structs
+function generic_projectto(x::T; kw...) where {T}
+    # Generic fallback, recursively make `ProjectTo`s for all their fields
     fields_nt::NamedTuple = backing(x)
-    return ProjectTo{T}(map(ProjectTo, fields_nt))
+    fields_proj = map(fields_nt) do x1
+        if x1 isa Number || x1 isa AbstractArray
+            ProjectTo(x1)
+        else
+            x1
+        end
+    end        
+    # We can't use `T` because if we have `Foo{Matrix{E}}` it should be allowed to make a
+    # `Foo{Diagaonal{E}}` etc. We assume it has a default constructor that has all fields 
+    # but if it doesn't `construct` will give a good error message.
+    wrapT = T.name.wrapper
+    return ProjectTo{wrapT}(; fields_proj..., kw...)
 end
 function (project::ProjectTo{T})(dx::Tangent) where {T}
     sub_projects = backing(project)
     sub_dxs = backing(canonicalize(dx))
-    _call(f, x) = f(x)
-    return construct(T, map(_call, sub_projects, sub_dxs))
-end
-
-# should not work for Tuples and NamedTuples, as not valid tangent types
-function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
-    return throw(
-        ArgumentError("The `x` in `ProjectTo(x)` must be a valid differential, not $x")
-    )
+    maybe_call(f::ProjectTo, x) = f(x)
+    maybe_call(f, x) = f
+    return construct(T, map(maybe_call, sub_projects, sub_dxs))
 end
 
 # Generic
+(::ProjectTo{T})(dx::T) where {T} = dx 
+(::ProjectTo{T})(dx::AbstractZero) where {T} = dx
+(::ProjectTo{T})(dx::NotImplemented) where {T} = dx
+
+ProjectTo() = ProjectTo{Any}()  # trivial case, exists so that maybe_call(f, x) knows what to do
+(x::ProjectTo{Any})(dx) = dx
+
+# Thunks
+(project::ProjectTo)(dx::Thunk) = Thunk(project ∘ dx.f)
+(project::ProjectTo)(dx::InplaceableThunk) = project(dx.val)  # can't update in-place part
 (project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
-(::ProjectTo{T})(dx::T) where {T} = dx  # not always true, but we can special case for when it isn't
-(::ProjectTo{T})(dx::AbstractZero) where {T} = zero(T)
 
-# Number
-ProjectTo(::T) where {T<:Number} = ProjectTo{T}()
+# Zero
+ProjectTo(::AbstractZero) = ProjectTo{AbstractZero}()
+(::ProjectTo{AbstractZero})(dx) = ZeroTangent()
```

Comment:
Are we sure we want to treat the two AbstractZeros the same here?
I think we want to treat each distinctly.
Because we do want that info in the output.


```suggestion
ProjectTo(::T) where T<:AbstractZero = ProjectTo{AbstractZero}()
(::ProjectTo{T})(dx) where T<:AbstractZero = T()
```

Alt: maybe we only want to define

```suggestion
ProjectTo(::NoTangent) = ProjectTo{AbstractZero}()
(::ProjectTo{NoTangent})(dx) = NoTangent()
```

Reply:
In the backward pass, all projectors should always let any any AbstractZero pass through, without change.

These lines are only treating the case that it (somehow?) ends up getting an AbstractZero in the forward pass. If it does, and then something nonzero arrives as gradient, that should be discarded for sure. I tried to write the simplest projector which would achieve that. Whether it should return `ZeroTangent()` or `NoTangent()` I have no opinion.

Doing this (Alt suggestion) won't work, I think, since nothing produces `ProjectTo{NoTangent}`:
```
ProjectTo(::NoTangent) = ProjectTo{AbstractZero}()
(::ProjectTo{NoTangent})(dx) = NoTangent()
```

The first suggestion broadens what I had so that particular AbstractZeros in the forward pass will, if they get nonzero gradient, re-create the same Zero type. But I'm not sure we need keep that much info.

Reply:
> In the backward pass, all projectors should always let any any AbstractZero pass through, without change. I 

That is a good point.
But consider a pullback that has been written for `sum(::AbstractArray{<:Real})`.
but is being invoked on a `Vector{Bool}`.
This means we initially do get a gradient of reals, but then out ProjectTo needs to cause us to end-up ant `NoTangent()`.
(or at least `Vector{NotTangent}()`)
Not `Vector{Zero}`



> Doing this (Alt suggestion) won't work, I think, since nothing produces `ProjectTo{NoTangent}`:

oops sorry typed it wrong
Meant to say:
```
ProjectTo(::NoTangent) = ProjectTo{NoTangent}()
(::ProjectTo{NoTangent})(dx) = NoTangent()
```

Which is the thing we need for `ProjectTo{Bool}` to get the right space.

Reply:
Yea ok. I can change which kind of Zero is returned, I didn't worry about any differences. And I can change what the projector which does that is called. (The same one gets used for Bool.)